sdk/core/src/policy.rs

```diff
@@ -118,30 +68,68 @@ impl Context {
 }
 
 #[async_trait::async_trait]
-pub trait Policy: Send + Sync + std::fmt::Debug {
-    async fn send(
-        &self,
-        ctx: Context,
-        request: &mut Request,
-        next: &[Arc<dyn Policy>],
-    ) -> PolicyResult<Response>;
+pub trait Policy<R>: Send + Sync + std::fmt::Debug
```

Comment:
I'm confused why `Policy` is generic...

Reply:
The idea is that a Policy can return either an updated Request (for example the CredentialsPolicy will add the authorization headers) or a Response. 

I was wondering if returning **always** a stream from TransportPolicy can be a good solution. We could then add an additional policy called `CollectStreamPolicy` that transforms the stream into a `bytes::Bytes`. Only functions that need to return the collected stream will add this additional policy. For example, for create database, the pipeline could be:

1. Prepare headers + method + body (specific of create database).
2. `CredentialsPolicy` adds auth headers (or bearer token).
3. `TelemetryPolicy` will add telemetry headers.
3. `TransportPolicy` *spends* the request and gives back a stream.
4. `CollectStreamPolicy` converts the stream into `bytes::Bytes`.
5. A custom `CreateDatabasePolicy` parses the stream and generates a `CreateDatabaseResponse`.

This way the entire create database code will be expressed in policies (most of which are reused across the crates). 

Does it make sense?

Reply:
Hmm.... The idea was that policies return the response. If the policy is receiving an owned `Request` why would it need to return the `Request`? This seems to complicate things. Perhaps you could show an example where this allows for flexibility where the non-generic trait could not handle.

Reply:
@MindFlavor I'm still not convinced that this architecture is the right way to go - having `Policy` be generic makes things more complicated than they need be.